File: Project2/lib/card.rb
# contains all classes and methods related to cards
module Cards
  # Class for Card Object
  class SetCard
    attr_reader :color, :number, :shape, :texture, :image_path
    attr_accessor :x, :y, :width, :height

    # Constructor Init
    def initialize(color, number, shape, texture, path)
      @color = color
      @number = number
      @shape = shape
      @texture = texture
      @image_path = path
      @x = -1
      @y = -1
      @width = 0
      @height = 0
    end

    # Creates an array of card image
    # Parameter card_deck: current deck of cards for game state
    # Parameter index: current place in deck to add card
    def create_card_sprite(card_deck, index)
      card_deck.displayed_sprites[index] = Sprite.new(
        @image_path.to_s,
        x: @x,
        y: @y
      )
    end

    # checks if card is clicked
    # Parameter event: GUI event to determine if a card was clicked
    # Returns true/false if a valid card is clicked/not
    def clicked?(event)
      (event.x > @x) && (event.x < @width) && (event.y > @y) && (event.y < @height)
    end

    # behavior of card when it is clicked on gui
    # Parameter card_deck: current deck of cards for game state
    # Parameter player: current player selected in the game
    def clicked_behavior(card_deck, player)
      # add the cards if not already selected
      card_deck.add_to_selected if (card_deck.selected_index.is_a? Integer) && (card_deck.index_array.length < 3)
      # check if length is 3 after adding
      if (card_deck.index_array.length == 3) && (card_deck.is_set? card_deck.cards_selected)
        puts 'Set Found'
        # replace cards
        card_deck.replace_card
        # increment score
        player.increment_score
        # clear selection after set is found
        card_deck.clear_selected
      elsif (card_deck.index_array.length == 3) && !(card_deck.is_set? card_deck.cards_selected)
        puts 'Sorry, not a set!'
        player.decrement_score
        # clear selection
        card_deck.clear_selected
      end
    end
  end

  # Function to Display all the cards on GUI (6 cards on each row)
  # Parameter card_deck: current deck of cards for game state
  def self.print_cards(card_deck)
    # Get number of rows needed
    num_rows = card_deck.current_deck.length / 6
    # Init x and y location on gui
    x_loc = 50
    y_loc = 50
    # Start index from position 0
    index = 0
    # Delete all the old sprite objects
    card_deck.displayed_sprites.each_index do |i|
      card_deck.displayed_sprites[i].remove
    end
    # Clear the sprite array
    card_deck.displayed_sprites.clear
    # For each row except the last one keep displaying 6 cards on each row
    num_rows.times do
      x_loc = 50
      6.times do
        # Get card and set new x and y location
        card = card_deck.current_deck[index]
        card.x = x_loc
        card.y = y_loc
        # Create sprite object to display card
        card.create_card_sprite(card_deck, index)
        index += 1
        x_loc += 110
      end
      # update y location each row
      y_loc += 56
    end
    # Set back to original x location
    x_loc = 50
    # Display the last cards on the last row
    (index...card_deck.current_deck.length).each do |pos|
      card = card_deck.current_deck[pos]
      card.x = x_loc
      card.y = y_loc
      card.create_card_sprite(card_deck, pos)
      x_loc += 110
    end
  end
end

File: lib/javascript/card.js
// class for card objects
class Cards {
  constructor(color, number, shape, texture, path) {
    this.color = color;
    this.number = number;
    this.shape = shape;
    this.texture = texture;
    this.path = path;
    this.selected = false;
  }

  //Add card to sprite array
  create_card_sprite(card_deck, index) {
    card_deck.displayed_cards[index] = this;
  }

  //Call Back function when card is clicked
  clicked(card_deck, elem) {
    if (this.selected) {
      this.remove_property(elem);
      this.remove_card(elem);
      this.selected = false;
    } else {
      this.keep_property(elem);
    
    card_deck.select_cards(elem);}
    console.log(card_deck.array_index);
    console.log(card_deck.cards_selected);
  }

  remove_card(elem) {
    card_deck.array_index.splice(card_deck.array_index.indexOf(elem.id - 1), 1);
    card_deck.cards_selected.splice(card_deck.cards_selected.indexOf(this), 1);
  }

  //Remove Border Property After displaying new cards
  remove_property(elem) {
    elem.classList.remove("card-selected");
  }

  //Keep Border Property if card is selected and gui has changed (for example when three new cards are added)
  keep_property(elem) {
    elem.classList.add("card-selected");
  }

  //Behavior of card When Clicked
  clicked_behavior() {
    if (!isNaN(card_deck.selected_index) && card_deck.array_index.length < 3) {
      card_deck.add_to_selected();
    }
    //Check if 3 cards are selected
    if (
      card_deck.array_index.length == 3 &&
      card_deck.is_set(card_deck.cards_selected)
    ) {
      window.alert("Set is Found!");
      //Replace with new cards
      card_deck.replace_card();
      game.current_player.increment_score();
      //Remove Border Selection Property
      for (let c of card_deck.cards_selected) {
        c.selected = false;
      }
      //Clear current cards selection
      card_deck.cards_selected = [];
      card_deck.array_index = [];
    } else if (
      card_deck.array_index.length == 3 &&
      !card_deck.is_set(card_deck.cards_selected)
    ) {
      window.alert("Not a Set!");
      //Decrement score
      game.current_player.decrement_score();
      for (let c of card_deck.cards_selected) {
        c.selected = false;
      }
      let cards_to_revert = document.getElementsByClassName("card-selected");
      for (let i = 0; i < 3; i++) {
        this.remove_property(cards_to_revert[0]); // Note: cards_to_revert shrinks as remove_property is called, so remove element at 0
      }

      //Clear current cards selection
      card_deck.cards_selected = [];
      card_deck.array_index = [];
    }
  }

  // display cards to html
  print_cards(card_deck) {
    //Get the number of rows that can display 6 cards
    let num_cards = Math.floor(card_deck.current_deck.length / 6);
    //Clear sprites array
    card_deck.displayed_cards = [];
    //Remove all child nodes of card container
    let elem = document.getElementById("card-container");
    this.remove_all_child();
    let index = 0;
    //Now create each row and display 6 cards on each row
    for (let i = 0; i < num_cards; i++) {
      index = this.recreate_row_of_cards(index);
    }
    //Display last row
    //Create html child for last row
    let r;
    if (index < card_deck.current_deck.length) {
      r = document.createElement("div");
      r.classList.add("card-container-row");
      elem.appendChild(r);
    }
    for (let k = index; k < card_deck.current_deck.length; k++) {
      this.create_card_img(index, r);
      index++;
    }
  }

  // removes all <img> from .card-container
  remove_all_child() {
    //Remove all child nodes of card container
    let elem = document.getElementById("card-container");
    let child = elem.lastElementChild;
    //remove all but the first child (prevent restart-game button from being removed)
    while (elem.childElementCount > 1 && child) {
      elem.removeChild(child);
      child = elem.lastElementChild;
    }
  }

  // creates a row of 6 cards inside .card-container
  recreate_row_of_cards(index) {
    let elem = document.getElementById("card-container");
    //Create html child for each row
    let r = document.createElement("div");
    r.classList.add("card-container-row");
    elem.appendChild(r);
    for (let j = 0; j < 6; j++) {
      this.create_card_img(index, r);
      index++;
    }
    return index;
  }

  // creates a <img> tag for card
  create_card_img(index, r) {
    let c_node = document.createElement("img");
    c_node.classList.add("card");
    c_node.id = `${index + 1}`;
    //Get the cards image path
    let curr_card = card_deck.current_deck[index];
    c_node.src = curr_card.path;

    //Add to DOM
    r.appendChild(c_node);
    //Add card to sprite array
    this.create_card_sprite(card_deck, index);
    if (curr_card.selected) {
      this.keep_property(c_node);
    }
    //register on click event
    c_node.addEventListener("click", () =>
      curr_card.clicked(card_deck, c_node)
    );
  }
}
